### ISSUE TITLE
The drawing of the similarity matrix and the t-SNE graph in the article

### ISSUE THREAD

**Issue by ABCBeacuse** [alexandor91/se3-equi-graph-registration#10]
Hello, your work is extremely excellent！！！，I am also very grateful for your patient answers to the questions I have raised over the years. It brought me, a beginner, a great inspiration.

Recently, I encountered some problems when writing and drawing the similarity matrix diagram and t-SNE diagram in similar articles. I hope to refer to your drawing code.

Could you please provide the drawing code conveniently? Thank you very much. I'm very sorry to disturb you again.

![Image](https://github.com/user-attachments/assets/3c7c96ce-f04b-4e12-843a-0bd19be58950)

![Image](https://github.com/user-attachments/assets/11ae34ab-2e07-4067-8765-546245f3c7ec)

**Comment by alexandor91** [comment]
Hi, thanks a lot for your interest to the work, actually we just use the common t-SNE and matplotlib library to plot 3D and 2D figures, here is the code snippet we used to plot, We first save the feature tensor output into npy files, then we load and visualize it offline.
`import numpy as np
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE

def load_features(file_path):
    """Load point features from numpy file (expected shape: N x 32)"""
    return np.load(file_path)

def visualize_tsne_3d(features, title="t-SNE Visualization"):
    """Visualize Nx32 features using t-SNE in 3D"""
    # Reduce dimensionality with t-SNE
    tsne = TSNE(n_components=3, random_state=42, perplexity=min(30, features.shape[0]-1))
    features_3d = tsne.fit_transform(features)
    
    # Create 3D plot
    fig = plt.figure(figsize=(10, 8))
    ax = fig.add_subplot(111, projection='3d')
    
    # Scatter plot with point indices as colors
    scatter = ax.scatter(features_3d[:, 0], features_3d[:, 1], features_3d[:, 2],
                         c=np.arange(features_3d.shape[0]), cmap='viridis', s=50)
    
    ax.set_title(f"{title} (N={features.shape[0]}, Original Dim=32)")
    ax.set_xlabel('x')
    ax.set_ylabel('y')
    ax.set_zlabel('z')
    plt.colorbar(scatter, ax=ax, label='Point Index')
    plt.tight_layout()
    plt.show()

def calculate_similarity(source_features, target_features):
    """Calculate normalized dot product similarity matrix between Nx32 features"""
    # Normalize features (L2 normalization)
    source_norm = source_features / (np.linalg.norm(source_features, axis=1, keepdims=True) + 1e-10)
    target_norm = target_features / (np.linalg.norm(target_features, axis=1, keepdims=True) + 1e-10)
    
    # Calculate cosine similarity matrix
    similarity = np.dot(source_norm, target_norm.T)
    return similarity

def visualize_similarity(similarity, source_name="Source", target_name="Target"):
    """Visualize NxN similarity matrix as heatmap"""
    fig, ax = plt.subplots(figsize=(10, 8))
    cax = ax.matshow(similarity, cmap='magma', vmin=0, vmax=1)
    
    ax.set_title(f"Feature Similarity: {source_name} vs {target_name}\n(Shape: {similarity.shape})")
    ax.set_xlabel(f'{target_name} Features')
    ax.set_ylabel(f'{source_name} Features')
    
    # Add colorbar with similarity range
    cbar = fig.colorbar(cax, ax=ax)
    cbar.set_label('Cosine Similarity', rotation=270, labelpad=15)
    
    plt.tight_layout()
    plt.show()

# Example usage
if __name__ == "__main__":
    # Load your features (replace with actual file paths)
    source_features = load_features("source_features.npy")  # N x 32
    target_features = load_features("target_features.npy")  # N x 32
    
    print(f"Source features shape: {source_features.shape}")
    print(f"Target features shape: {target_features.shape}")
    
    # Visualize features separately
    visualize_tsne_3d(source_features, "Source Features")
    visualize_tsne_3d(target_features, "Target Features")
    
    # Calculate and visualize similarity matrix
    similarity = calculate_similarity(source_features, target_features)
    visualize_similarity(similarity, "Source", "Target")`
